File: src/main/java/com/hileco/cortex/analysis/processors/JumpIllegalProcessor.java
package com.hileco.cortex.analysis.processors;

import com.hileco.cortex.analysis.Graph;
import com.hileco.cortex.analysis.edges.EdgeFlowMapping;
import com.hileco.cortex.instructions.debug.HALT;
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION;

import java.util.stream.Collectors;

import static com.hileco.cortex.analysis.edges.EdgeFlowType.INSTRUCTION_JUMP;
import static com.hileco.cortex.instructions.ProgramException.Reason.JUMP_OUT_OF_BOUNDS;
import static com.hileco.cortex.instructions.ProgramException.Reason.JUMP_TO_ILLEGAL_INSTRUCTION;

public class JumpIllegalProcessor implements Processor {

    @Override
    public void process(Graph graph) {
        EdgeFlowMapping.UTIL.findAny(graph).ifPresent(edgeFlowMapping -> edgeFlowMapping.getFlowsFromSource().forEach((sourceLine, edgeFlows) -> {
            var jumpEdgeFlows = edgeFlows.stream()
                    .filter(edgeFlow -> edgeFlow.getType() == INSTRUCTION_JUMP)
                    .collect(Collectors.toSet());
            if (jumpEdgeFlows.size() == 1) {
                var onlyJumpEdgeFlow = jumpEdgeFlows.stream().findFirst().orElseThrow();
                var targetedNode = edgeFlowMapping.getNodeLineMapping().get(onlyJumpEdgeFlow.getTarget());
                var sourceNode = edgeFlowMapping.getNodeLineMapping().get(onlyJumpEdgeFlow.getSource());
                if (targetedNode == null) {
                    sourceNode.getInstruction().set(new HALT(JUMP_OUT_OF_BOUNDS));
                } else if (!(targetedNode.getInstruction().get() instanceof JUMP_DESTINATION)) {
                    sourceNode.getInstruction().set(new HALT(JUMP_TO_ILLEGAL_INSTRUCTION));
                }
            }
        }));
    }
}

File: src/main/java/com/hileco/cortex/analysis/processors/JumpIllegalProcessor.kt
package com.hileco.cortex.analysis.processors

import com.hileco.cortex.analysis.Graph
import com.hileco.cortex.analysis.edges.FlowMapping
import com.hileco.cortex.analysis.edges.FlowType.INSTRUCTION_JUMP
import com.hileco.cortex.instructions.ProgramException.Reason.JUMP_TO_ILLEGAL_INSTRUCTION
import com.hileco.cortex.instructions.ProgramException.Reason.JUMP_TO_OUT_OF_BOUNDS
import com.hileco.cortex.instructions.debug.HALT
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION

class JumpIllegalProcessor : Processor {
    override fun process(graph: Graph) {
        val flowMapping = graph.edgeMapping.get(FlowMapping::class.java).single()
        flowMapping.flowsFromSource.values.forEach { flows ->
            val onlyJumpFlow = flows.singleOrNull { flow -> flow.type == INSTRUCTION_JUMP }
            if (onlyJumpFlow != null) {
                val sourceNode = flowMapping.nodeLineMapping[onlyJumpFlow.source]
                if (sourceNode != null) {
                    val targetedNode = flowMapping.nodeLineMapping[onlyJumpFlow.target]
                    if (targetedNode == null) {
                        sourceNode.instruction = HALT(JUMP_TO_OUT_OF_BOUNDS)
                    } else if (targetedNode.instruction !is JUMP_DESTINATION) {
                        sourceNode.instruction = HALT(JUMP_TO_ILLEGAL_INSTRUCTION)
                    }
                }
            }
        }
    }
}
